Add docker registry handling to master

Cynerva opened the pull request:
This makes Calico, Canal, or Tigera Secure EE work with the docker-registry charm, by making it possible to relate kubernetes-master to docker-registry. This was done by moving the layer-docker and registry handling code from kubernetes-worker to kubernetes-common.

There is an architectural change happening here: kubernetes-common, which previously only contained a library of functions for any charm to use, now has its own relations and reactive handlers. So it now represents the common reactive code between kubernetes-master and kubernetes-worker, and can no longer be included in other charms without side effects. As a consequence, I had to remove layer:kubernetes-common from kubeapi-load-balancer.

I think this is a step in the right direction. It seems like kubernetes-master and kubernetes-worker tend to converge slowly over time toward the same point. Where kube-proxy was previously only run on workers, it now runs on masters too. Where docker handling was previously only done by worker, it now has to be done by masters too. We have discussed some day running kubelet on masters as well.

If we architect kubernetes-common specifically as a base layer for kubernetes-master and kubernetes-worker, then kubernetes-common can be the place where that convergence happens.

Thoughts and opinions welcome.

Cynerva commented:
@kwmonroe @hyperbolic2346 I would like both of your opinions on the architectural change to kubernetes-common happening here.

johnsca commented:
I'm generally ok with this but wonder if: 1) the name should maybe be changed, since it's no longer truly common (to all of the k8s charms), 2) if there would be some utility in having an actual truly common layer to avoid duplication of helpers (or perhaps a k8s_charms_helpers lib or something)?

Cynerva commented:
Per IRC, the plan is to leave layer-kubernetes-common as a lib layer with no side effects, and introduce a new layer: layer-kubernetes-master-worker-base. Working on that now.

Cynerva commented:
Done, I've reverted changes to layer-kubernetes-common, and added layer-kubernetes-master-worker-base instead.

This now depends on https://github.com/juju/layer-index/pull/62

hyperbolic2346 commented on cluster/juju/layers/kubernetes-master-worker-base/reactive/kubernetes_master_worker_base.py:
```diff
@@ -0,0 +1,141 @@
+
+
+@when('endpoint.docker-registry.ready')
+@when_not('kubernetes-master-worker-base.registry.configured')
```
that's a mouthful :)

hyperbolic2346 commented on cluster/juju/layers/kubernetes-master/layer.yaml:
```diff
@@ -8,7 +8,7 @@ includes:
   - 'layer:metrics'
   - 'layer:nagios'
   - 'layer:cdk-service-kicker'
-  - 'layer:kubernetes-common'
```
master and worker are no longer using kubernetes-common?

Cynerva replied:
They're inheriting it via kubernetes-master-worker-base. Do you want me to explicitly include it in kubernetes-master and kubernetes-worker too?

kwmonroe replied:
-1 on being explicit, but it's a soft -1.  We don't explicitly list layers that get inherited from things like `layer:vaultlocker`.

That said, i could see how one might glance at the code and wonder how we're including common functions if the layer isn't listed, and then having to take a minute to figure out where they must be coming from...

So this comment is basically worthless. I'm confident you'll do the right thing.

johnsca replied:
-1 on being explicit as well, since the point of the base layer is to avoid repeating things that are shared.

On a more general note, by philosophy is to be explicit about the include if the layer is explicitly referenced in the charm, unless it's being included by something that is an explicit base layer for a small set of related charms (i.e., not a general layer used by many unrelated charms).  If that makes any sense.
